### PR TITLE
snat_translation module

### PR DESCRIPTION
**Pull request by crosbygw**
Added snat_translation module based on feedback from github issue: https://github.com/F5Networks/f5-ansible/issues/1227

### CODE REVIEW

**Comment by wojtek0806** [comment]
squash and rebase before i even attempt to review please

File: library/modules/bigip_snat_translation.py [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+
+
+class ArgumentSpec(object):
+    def __init__(self):
+        self.supports_check_mode = True
+        argument_spec = dict(
+            address=dict(
+                aliases=['address', 'ip'],
+            ),
+            arp=dict(
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: this is boolean, also remove defaults, we do not provide defaults for API parameters where bigip fills in the blanks, see Chapter 3 of the developer guide.

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_firewall_log_profile.py#L851

File: library/modules/bigip_snat_translation.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+            ),
+            state=dict(
+                default='present',
+                choices=['absent', 'present', 'enabled', 'disabled']
+            ),
+            tcp_idle_timeout=dict(
+                default='indefinite',
+            ),
+            traffic_group=dict(
+                default='/Common/traffic-group-1',
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove default

File: library/modules/bigip_snat_translation.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+            ),
+            udp_idle_timeout=dict(
+                default='indefinite',
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove default

File: library/modules/bigip_snat_translation.py [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+                default='enabled',
+                choices=['enabled', 'disabled']
+            ),
+            connection_limit=dict(
+                type='int',
+                default='0',
+            ),
+            description=dict(),
+            ip_idle_timeout=dict(
+                default='indefinite',
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove default

File: library/modules/bigip_snat_translation.py [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+            address=dict(
+                aliases=['address', 'ip'],
+            ),
+            arp=dict(
+                default='enabled',
+                choices=['enabled', 'disabled']
+            ),
+            connection_limit=dict(
+                type='int',
+                default='0',
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+
+class ArgumentSpec(object):
+    def __init__(self):
+        self.supports_check_mode = True
+        argument_spec = dict(
+            address=dict(
+                aliases=['address', 'ip'],
+            ),
+            arp=dict(
+                default='enabled',
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: no defaults

File: library/modules/bigip_snat_translation.py [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+                raise F5ModuleError(resp.content)
+        return ApiParameters(params=response)
+
+
+class ArgumentSpec(object):
+    def __init__(self):
+        self.supports_check_mode = True
+        argument_spec = dict(
+            address=dict(
+                aliases=['address', 'ip'],
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: why do you have address twice ? if alias then ip is only alias needed

File: library/modules/bigip_snat_translation.py [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+      - Specifies the IP address of the SNAT translation. When a C(state) of present, enabled, or disabled is
+        provided, this parameter is required. When c(absent), it is optional. This parameter cannot be updated after it is set.
+    type: str
+    aliases:
+      - address
+      - ip
+  arp:
+    description:
+      - Specifies, when enabled, that the NAT sends ARP requests. When disabled, does not send ARP requests.
+      type: str
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: this should be boolean

File: library/modules/bigip_snat_translation.py [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+      choices:
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+        - enabled
+        - disabled
+      default: enabled
+  connection_limit:
+    description:
+      - Specifies a limit on the number of connections a translation address must reach before it no longer
+        initiates a connection. The default value of 0 indicates that the setting is disabled.
+      type: int
+      default: 0
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: no defaults remove

File: library/modules/bigip_snat_translation.py [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+  description:
+    description:
+      - Description of snat-translation. c(none) will set to default description of null.
+      type: str
+      default none
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: no defaults remove

File: library/modules/bigip_snat_translation.py [hunk 12]
```diff
@@ -0,0 +1,730 @@
...
+  ip_idle_timeout:
+    description:
+      - Specifies the amount of time that connections to an IP address initiated using a SNAT address are
+        allowed to remain idle before being automatically disconnected. C(indefinite) prevents the connection
+        from timing out.
+      - The accepted value range is C(0 - 4294967295) seconds, specifying C(indefinite) will
+        set it to the maximum value.
+    type: str
+    default: indefinite
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+    default: present
+  tcp_idle_timeout:
+    description:
+      - Specifies the amount of time that TCP connections initiated using a SNAT address are allowed
+        to remain idle before being automatically disconnected. C(indefinite) Prevents the
+        connection from timing out.
+      - The accepted value range is C(0 - 4294967295) seconds, specifying C(indefinite) will
+        set it to the maximum value.
+    type: str
+    default: indefinite
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 14]
```diff
@@ -0,0 +1,730 @@
...
+  traffic_group:
+    description:
+      - The traffic group for the snat-translation address. When creating a new address,
+        if this value is not specified, the default of C(/Common/traffic-group-1)
+        will be used.
+    type: str
+    default: /Common/traffic-group-1
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 15]
```diff
@@ -0,0 +1,730 @@
...
+  udp_idle_timeout:
+    description:
+      - Specifies the amount of time that UDP connections initiated using a SNAT address are allowed
+        to remain idle before being automatically disconnected. C(indefinite) Prevents the connection
+        from timing out.
+      - The accepted value range is C(0 - 4294967295) seconds, specifying C(indefinite) will
+        set it to the maximum value.
+    type: str
+    default: indefinite
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove

File: library/modules/bigip_snat_translation.py [hunk 16]
```diff
@@ -0,0 +1,730 @@
...
+      user: admin
+      password: secret
+  delegate_to: localhost
+
+- name: Disable a SNAT translation 'my-snat-translation'
+  bigip_snat_translation:
+    name: my-snat-pool
+    state: disabled
+    address: 10.10.10.10
+    arp: diabled
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: change

File: library/modules/bigip_snat_translation.py [hunk 17]
```diff
@@ -0,0 +1,730 @@
...
+      user: admin
+      password: secret
+  delegate_to: localhost
+
+- name: Enable a SNAT translation 'my-snat-translation'
+  bigip_snat_translation:
+    name: my-snat-pool
+    state: enabled
+    address: 10.10.10.10
+    arp: diabled
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: change

File: library/modules/bigip_snat_translation.py [hunk 18]
```diff
@@ -0,0 +1,730 @@
...
+      user: admin
+      password: secret
+  delegate_to: localhost
+
+- name: Create using partition other then /Common on a SNAT translation 'my-new-snat-translation'
+  bigip_snat_translation:
+    name: my-new-snat-pool
+    state: enabled
+    address: 10.10.10.10
+    arp: diabled
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: change

File: library/modules/bigip_snat_translation.py [hunk 19]
```diff
@@ -0,0 +1,730 @@
...
+      user: admin
+      password: secret
+  delegate_to: localhost
+
+- name: Modify using traffic group other then /Common/traffic-group-1 on a SNAT translation 'my-new-snat-translation'
+  bigip_snat_translation:
+    name: my-new-snat-pool
+    state: enabled
+    address: 10.10.10.10
+    arp: diabled
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: change

File: library/modules/bigip_snat_translation.py [hunk 20]
```diff
@@ -0,0 +1,730 @@
...
+'''
+
+RETURN = r'''
+address:
+  description:
+    - ip address used for SNAT translation.
+  returned: changed and success
+  type: str
+  sample: "10.10.10.10"
+arp:
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: change accordingly

File: library/modules/bigip_snat_translation.py [hunk 21]
```diff
@@ -0,0 +1,730 @@
...
+        'udp_idle_timeout'
+    ]
+
+    @property
+    def traffic_group(self):
+        if self._values['traffic_group'] is None:
+            return None
+        else:
+            result = fq_name(self.partition, self._values['traffic_group'])
+        if result.startswith('/Common/'):
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: this is not needed fq_name takes care of this, change it

File: library/modules/bigip_snat_translation.py [hunk 22]
```diff
@@ -0,0 +1,730 @@
...
+            return result
+        else:
+            raise F5ModuleError(
+                "Traffic groups can only exist in /Common"
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: This is not needed as this is a well known fact, you can safely remove this

File: library/modules/bigip_snat_translation.py [hunk 23]
```diff
@@ -0,0 +1,730 @@
...
+
+    @property
+    def connection_limit(self):
+        if self._values['connection_limit'] is None:
+            return None
+        return int(self._values['connection_limit'])
+
+    @property
+    def description(self):
+        result = cmp_str_with_none(self.want.description, self.have.description)
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: what are you comparing here ? Difference class is where you want this

File: library/modules/bigip_snat_translation.py [hunk 24]
```diff
@@ -0,0 +1,730 @@
...
+        attr1 = getattr(self.want, param)
+        try:
+            attr2 = getattr(self.have, param)
+            if attr1 != attr2:
+                return attr1
+        except AttributeError:
+            return attr1
+
+    @property
+    def description(self):
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove and replace with cmp_str_with_none function

File: library/modules/bigip_snat_translation.py [hunk 25]
```diff
@@ -0,0 +1,730 @@
...
+        if self._values['description'] is None:
+            return None
+        elif self._values['description'] in ['none', '']:
+            return ''
+        return self._values['description']
+
+    @property
+    def traffic_group(self):
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: not needed here this type of comparison is taken care of by the compare method

File: library/modules/bigip_snat_translation.py [hunk 26]
```diff
@@ -0,0 +1,730 @@
...
+  returned: changed
+  type: bool
+  sample: True
+ip_idle_timeout:
+  description: IP idle timeout value for snat-translation.
+  returned: changed
+  type: str
+  sample: 300
+state:
+  description: The new state of the virtual address.
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: wrong description

File: library/modules/bigip_snat_translation.py [hunk 27]
```diff
@@ -0,0 +1,730 @@
...
+  description: Description of the snat-translaiton.
+  returned: changed
+  type: str
+  sample: My snat-translation
+disabled:
+  description: Whether the snat-translation is disabled, or not.
+  returned: changed
+  type: bool
+  sample: True
+enabled:
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove this as user did not provide this

File: library/modules/bigip_snat_translation.py [hunk 28]
```diff
@@ -0,0 +1,730 @@
...
+  description: The new connection limit of the virtual address.
+  returned: changed
+  type: int
+  sample: 1000
+description:
+  description: Description of the snat-translaiton.
+  returned: changed
+  type: str
+  sample: My snat-translation
+disabled:
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: remove as user has not provided this

File: library/modules/bigip_snat_translation.py (hunk 1 above)
Comment (crosbygw) [reply library/modules/bigip_snat_translation.py]: Regarding note on argument type, allowed/returned values are string - either "enabled" or "disabled" rather then boolean values: 0, 'on', 'f', 'false', 1, 'no', 'n', '1', '0', 't', 'y', 'off', 'yes', 'true'.

Comment (wojtek0806) [reply library/modules/bigip_snat_translation.py]: no the pattern is to have these converted to booleans, ansible params != rest params:

latest example:
https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_message_routing_peer.py#L272

File: library/modules/bigip_snat_translation.py [hunk 29]
```diff
@@ -0,0 +1,694 @@
...
+
+    @property
+    def connection_limit(self):
+        if self._values['connection_limit'] is None:
+            return None
+        return int(self._values['connection_limit'])
+
+    @property
+    def description(self):
+        if self._values['description'] in [None, 'none']:
```
[the remark this answers is not in the document]
Comment (wojtek0806) [reply library/modules/bigip_snat_translation.py]: Like here: https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_monitor_http.py#L378

File: library/modules/bigip_snat_translation.py [hunk 30]
```diff
@@ -0,0 +1,694 @@
...
+
+    @property
+    def state(self):
+        if self.enabled is True and self._values['state'] != 'present':
+            return 'enabled'
+        elif self.disabled is True:
+            return 'disabled'
+        else:
+            return self._values['state']
+
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: You are missing integer validation for:  connection_limit, ip_idle_timeout, udp_idle_timeout, udp_idle_timeout, connection_limit

something similar to this:

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_firewall_log_profile.py#L369

File: library/modules/bigip_snat_translation.py [hunk 31]
```diff
@@ -0,0 +1,694 @@
...
+            pass
+        return result
+
+
+class UsableChanges(Changes):
+    pass
+
+
+class ReportableChanges(Changes):
+    pass
```
Comment (wojtek0806) [inline library/modules/bigip_snat_translation.py]: are the returned values the same as user input them? you would want to do something like this if they are not, obviously not the same code but this is to get you the idea:
https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_message_routing_router.py#L427

File: test/integration/targets/bigip_snat_translation/tasks/main.yaml [hunk 32]
```diff
@@ -0,0 +1,243 @@
...
+    tcp_idle_timeout: "{{ st_tcp_idle_timeout }}"
+    udp_idle_timeout: "{{ st_udp_idle_timeout }}"
+  register: result
+
+- name: Assert Create SNAT translation using defaults
+  assert:
+    that:
+      - result is changed
+      - result is success
+      - result.address == "{{ st_address }}"
```
Comment (wojtek0806) [inline test/integration/targets/bigip_snat_translation/tasks/main.yaml]: when you do asserts the variables should be written without curly braces, e.g. 
`result.address == st_address` and so on

Comment (wojtek0806) [reply test/integration/targets/bigip_snat_translation/tasks/main.yaml]: make the above corrections for all the tests

**Comment by crosbygw** [comment]
unit and integration tests logs attached.

[test results.zip](https://github.com/F5Networks/f5-ansible/files/3337004/test.results.zip)

**Pull request merged**